### PR TITLE
Improve generation of thumbnails for files

### PR DESCRIPTION
**Pull request by jordisala1991**
## Subject

I am targeting this branch, because this breaks BC.

## Changelog

```markdown
### Added
- Added thumbnail placeholder generator for files.
- Split thumbnail interface to control wether they are generated or not.
```

### CODE REVIEW

File: src/Thumbnail/FileThumbnail.php [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    }
+
+    public function generate(MediaProviderInterface $provider, MediaInterface $media): void
+    {
+        // nothing to generate, as the thumbnails for files are already generated
+    }
+
+    public function delete(MediaProviderInterface $provider, MediaInterface $media, $formats = null): void
+    {
+        // feature not available
```
Comment (VincentLanglet) [inline src/Thumbnail/FileThumbnail.php:56]: Should we throw an exception ?

Comment (jordisala1991) [reply src/Thumbnail/FileThumbnail.php:56]: I copied another thumbnail implementation. Should I change it too?

Comment (VincentLanglet) [reply src/Thumbnail/FileThumbnail.php:56]: Honestly I dunno if we should throw a method not implemented exception or doing nothing becase some manager could use the delete method every time and rely on the fact the feature is really removed on not.

I would say the best solution would be to remove the `delete` method from the interface and add a DeletableThumbnailInterface for this. This way you can remove the delete method here.

Do you want to do this here, in another PR or never ? ^^

Comment (jordisala1991) [reply src/Thumbnail/FileThumbnail.php]: I tried to do it in this PR, wdyt?

File: src/Provider/BaseProvider.php [hunk 2]
```diff
@@ -153,12 +154,16 @@ public function requireThumbnails(): bool
 
     public function generateThumbnails(MediaInterface $media): void
     {
-        $this->thumbnail->generate($this, $media);
+        if ($this->thumbnail instanceof GeneratableThumbnailInterface) {
```
Comment (VincentLanglet) [inline src/Provider/BaseProvider.php:157]: I don't know enough this code to know if we have to do it this way or throw an exception instead. Do you know ?

File: src/Provider/BaseProvider.php [hunk 3]
```diff
@@ -153,12 +154,16 @@ public function requireThumbnails(): bool
...
+            $this->thumbnail->generate($this, $media);
+        }
     }
 
     public function removeThumbnails(MediaInterface $media, $formats = null): void
     {
-        $this->thumbnail->delete($this, $media, $formats);
+        if ($this->thumbnail instanceof GeneratableThumbnailInterface) {
```
Comment (VincentLanglet) [inline src/Provider/BaseProvider.php:164]: Same

File: src/Thumbnail/ConsumerThumbnail.php [hunk 4]
```diff
@@ -60,12 +60,13 @@ public function generatePrivateUrl(MediaProviderInterface $provider, MediaInterf
 
     public function generate(MediaProviderInterface $provider, MediaInterface $media): void
     {
-        $backend = $this->backend;
-        $id = $this->id;
+        if (!$this->thumbnail instanceof GeneratableThumbnailInterface) {
```
Comment (VincentLanglet) [inline src/Thumbnail/ConsumerThumbnail.php:63]: Same

File: src/Thumbnail/ConsumerThumbnail.php [hunk 5]
```diff
@@ -80,6 +81,8 @@ public function generate(MediaProviderInterface $provider, MediaInterface $media
 
     public function delete(MediaProviderInterface $provider, MediaInterface $media, $formats = null): void
     {
-        $this->thumbnail->delete($provider, $media, $formats);
+        if ($this->thumbnail instanceof GeneratableThumbnailInterface) {
```
Comment (VincentLanglet) [inline src/Thumbnail/ConsumerThumbnail.php:84]: Same

File: src/Thumbnail/FileThumbnail.php [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    public function __construct(Packages $packages)
+    {
+        $this->packages = $packages;
+    }
+
+    public function generatePublicUrl(MediaProviderInterface $provider, MediaInterface $media, string $format): string
+    {
+        if (MediaProviderInterface::FORMAT_ADMIN !== $format) {
+            throw new \RuntimeException(sprintf('Unable to generate thumbnail for the given format %s.', $format));
```
Comment (VincentLanglet) [inline src/Thumbnail/FileThumbnail.php:38]: Shouldn't it be an `InvalidArgumentException` or something like `Unsupported format` ?

Comment (VincentLanglet) [reply src/Thumbnail/FileThumbnail.php:38]: And `InvalidArgumentException` might be the behavior of other Thumbmail instead of returning `''`, for instance in `FormatThumbnail`, when the id is null.

File: src/Thumbnail/FileThumbnail.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        }
+
+        return $this->packages->getUrl('bundles/sonatamedia/file.png');
+    }
+
+    public function generatePrivateUrl(MediaProviderInterface $provider, MediaInterface $media, string $format): string
+    {
+        throw new \RuntimeException('Unable to generate private thumbnail url for media files.');
```
Comment (VincentLanglet) [inline src/Thumbnail/FileThumbnail.php:46]: BadMethodCallException maybe ?

I consider that if we throw a RuntimeException it must be added to the interface when it's not mandatory for LogicException. (And it's more a logic exception to call a method which is not supposed to be called)

File: tests/Provider/FileProviderTest.php [hunk 8]
```diff
@@ -43,8 +43,12 @@ class FileProviderTest extends AbstractProviderTest
     public function getProvider(): MediaProviderInterface
     {
         $resizer = $this->createMock(ResizerInterface::class);
+        $thumbnail = $this->createMock(ThumbnailInterface::class);
+        $metadata = $this->createMock(MetadataBuilderInterface::class);
```
Comment (phansys) [inline tests/Provider/FileProviderTest.php:47]: This seems a stub, at least in the context of this method.

File: tests/Thumbnail/FileThumbnailTest.php [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+namespace Sonata\MediaBundle\Tests\Security;
+
+use PHPUnit\Framework\MockObject\MockObject;
+use PHPUnit\Framework\TestCase;
+use Sonata\MediaBundle\Provider\FileProviderInterface;
+use Sonata\MediaBundle\Tests\App\Entity\Media;
+use Sonata\MediaBundle\Thumbnail\FileThumbnail;
+use Symfony\Component\Asset\Packages;
+
+class FileThumbnailTest extends TestCase
```
Comment (phansys) [inline tests/Thumbnail/FileThumbnailTest.php:23]:
```suggestion
final class FileThumbnailTest extends TestCase
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: src/Thumbnail/FileThumbnail.php [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+namespace Sonata\MediaBundle\Thumbnail;
+
+use Sonata\MediaBundle\Model\MediaInterface;
+use Sonata\MediaBundle\Provider\MediaProviderInterface;
+use Symfony\Component\Asset\Packages;
+
+/**
+ * @author Jordi Sala Morales <jordism91@gmail.com>
+ */
+final class FileThumbnail implements ThumbnailInterface
```
Comment (phansys) [inline src/Thumbnail/FileThumbnail.php:23]: The namespace already includes the term "Thumbnail".
IMO, "File" is enough for this class.

Comment (VincentLanglet) [reply src/Thumbnail/FileThumbnail.php:23]: I don't agree with this, or the interface could have been called `Interface` since it's in the Thumbnail namespace.
It avoids conflict with another File class and helps to give context when using the class.
The `ArrayCollection` is not called `Array` and there is lot of similar example.

File: src/Provider/BaseProvider.php (hunk 2 above)
Comment (jordisala1991) [reply src/Provider/BaseProvider.php:157]: I think it is better this way. This code could be called by subclasses or this own class, for each media saved.

File: src/Thumbnail/FileThumbnail.php (hunk 10 above)
Comment (jordisala1991) [reply src/Thumbnail/FileThumbnail.php:23]: All the code on this bundle and other sonata bundle are the same way. On this same folder for example, all other Thumbnails are called that way. Or you can check for Providers, all include the folder name again.

This is for consistency, but also I don't agree calling it simply "File", it could be misleading, it is not a file.

**Comment by jordisala1991** [comment]
Applied reviews.

**Review by VincentLanglet — APPROVED** [review]

File: src/Thumbnail/FileThumbnail.php [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+
+    public function __construct(Packages $packages)
+    {
+        $this->packages = $packages;
+    }
+
+    public function generatePublicUrl(MediaProviderInterface $provider, MediaInterface $media, string $format): string
+    {
+        if (MediaProviderInterface::FORMAT_ADMIN !== $format) {
+            throw new \InvalidArgumentException(sprintf('Unsupported format %s.', $format));
```
Comment (phansys) [inline src/Thumbnail/FileThumbnail.php:38]:
```suggestion
            throw new \InvalidArgumentException(sprintf('Unsupported format "%s".', $format));
```

File: src/Thumbnail/GeneratableThumbnailInterface.php [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace Sonata\MediaBundle\Thumbnail;
+
+use Sonata\MediaBundle\Model\MediaInterface;
+use Sonata\MediaBundle\Provider\MediaProviderInterface;
+
+/**
+ * @author Jordi Sala Morales <jordism91@gmail.com>
+ */
+interface GeneratableThumbnailInterface
```
Comment (phansys) [inline src/Thumbnail/GeneratableThumbnailInterface.php:22]: I don't know how correct it can be, but I think "Generable" fits better than "Generatable" in this case.
BTW, with a too specific name like the one this interface currently has, the `delete()` method seems unrelated to its responsibilities. IMO, the case could be different if the name were `Thumbnail(Manager|Handler)Interface` or similar.

Comment (jordisala1991) [reply src/Thumbnail/GeneratableThumbnailInterface.php]: Renamed to Generable. I kinda don't like super generics names because we will end up with more methods there instead of more interfaces.

I think about it this way: Thumbnails that are generated should be also deleted at some point (when you remove the media). There are the File Thumbnails, that are for files and can't be generated, so they cannot be removed either. And the last case if for Liip thumbnails, generated on demand and stored on some cache folder (no need to delete them because they will be removed on cache removal).

wdyt?

**Review by VincentLanglet — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

File: src/Thumbnail/GeneratableThumbnailInterface.php (hunk 12 above)
Comment (phansys) [reply src/Thumbnail/GeneratableThumbnailInterface.php]: It's fine for me. I just want to point out that this level of specificity might require a `DeletableThumbnailInterface`.
But I don't want to block this PR.

**Review by phansys — APPROVED** [review]

**Comment by jordisala1991** [comment]
Lets go with this implementation for now. the bundle needs more changes so we are still in time to change the interface again
